tsdb/engine/tsm1/cache.go

```diff
@@ -187,6 +189,15 @@ func (c *Cache) Snapshot() *Cache {
 		c.snapshotSize += uint64(Values(e.values).Size())
 	}
 
+	// Do deduplication a copy of the array.
```

Comment:
BTW: I agree that there is a simplification available, but I have not done that refactoring yet because it aids explanation of the solution if those two loops are kept separate. Once it is accepted that c.snapshot.dirty is required, I can do that refactoring.